forest/redux.py

```diff
@@ -99,18 +74,29 @@ class Store(Observable):
     def __init__(self, reducer, initial_state=None, middlewares=None):
         self.reducer = reducer
         self.state = initial_state if initial_state is not None else {}
-        if middlewares is not None:
-            mws = [m(self) for m in middlewares]
-            f = self.dispatch
-            for mw in reversed(mws):
-                f = mw(f)
-            self.dispatch = f
+        self.middlewares = middlewares if middlewares is not None else []
         super().__init__()
 
     def dispatch(self, action):
         """Apply reducer and notify listeners of new state
 
         :param action: plain dict consumed by the reducer
         """
-        self.state = self.reducer(self.state, action)
-        self.notify(self.state)
+        actions = self.pure(action)
+        for middleware in self.middlewares:
+            actions = self.bind(middleware, self, actions)
+        for _action in actions:
+            self.state = self.reducer(self.state, _action)
+            self.notify(self.state)
+
+    @staticmethod
+    def pure(action):
+        """Embed action into action generator"""
+        yield action
+
+    @staticmethod
+    def bind(middleware, store, actions):
+        """Flat map action generators from middleware into action generator"""
+        for action in actions:
+            for _action in middleware(store, action):
+                yield _action
```

Comment:
We should replace this with `yield from middleware(store, action)`.

Reply:
Agreed, a much cleaner syntax with no temporary variable